src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
+
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, findNodeByIndex(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            linkLast(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexElementException(index);
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexElementException(index);
+        Node<T> oldNode = findNodeByIndex(index);
+        T oldValue = oldNode.value;
+        oldNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexElementException(index);
+        return unlink(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        if (object == null) {
+            while (current != null) {
+                if (current.value == null) {
+                    unlink(current);
+                    return true;
+                }
+                current = current.next;
+            }
+        } else {
+            while (current != null) {
+                if (current.value != null && current.value.equals(object)) {
+                    unlink(current);
+                    return true;
+                }
+                current = current.next;
+            }
+        }
+
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkLast(T value) {
+        Node<T> last = tail;
+        Node<T> newNode = new Node<>(null, value, last);
+        tail = newNode;
+
+        if (last == null) {
+            head = newNode;
+        } else {
+            last.next = newNode;
+        }
+
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> current) {
+        Node<T> prev = current.prev;
+        Node<T> newNode = new Node<>(current, value, prev);
+        current.prev = newNode;
+
+        if (prev == null) {
+            head = newNode;
+        } else {
+            prev.next = newNode;
+        }
+
+        size++;
+    }
+    
+    private T unlink(Node<T> current) {
+        if (current == null) {
+            return null;
+        }
+
+        Node<T> next = current.next;
+        Node<T> prev = current.prev;
+
+        if (next == null) {
+            tail = prev;
+        }
+
+        if (prev == null) {
+            head = next;
+        }
+        
+        if (next != null && prev != null) {
+            next.prev = prev;
+            prev.next = next;
+        }
+
+        size--;
+        return current.value;
+    }
+    
+    private Node<T> findNodeByIndex(int index) {
```

Comment:
The current implementation iterates from the `head` node to the desired `index`. If the desired `index` is closer to the end of the list, it would be more efficient to start from the `tail` and move backward.